### PR TITLE
add a 'rank tutorials' helper and oneoff script [WIP]

### PR DESCRIPTION
**Pull request by Hamms**
for manual popularity ranking of HoC tutorials

### CODE REVIEW

**Comment by Hamms** [comment]
@ashercodeorg what do you think about this general approach?

File: lib/analyze_hoc_activity_helper.rb [hunk 1]
```diff
@@ -9,23 +9,78 @@
...
 
+WEIGHTED_COUNT = "SUM(" \
+  " IF(session REGEXP '^_.*_'," +
+  # Parses the session weight 'xyz' from the session '_xyz_abcdefghijk'.
+  "   SUBSTRING(SUBSTRING_INDEX(session, '_', 2) FROM 2)," +
+  # The session does not specify the session weight, default to one.
+  "   1)" \
+  " ) AS count"
+
+def generate_from_where(start_date, end_date: nil, tutorial: true, tutorial_pixel: true, finished: false)
```
Comment (ashercodeorg) [inline lib/analyze_hoc_activity_helper.rb:33]: nit: Please document that `start_date` is inclusive and `end_date` is exclusive.

Better, document the meaning of all parameters in a comment, as none are crystal clear.

File: lib/analyze_hoc_activity_helper.rb [hunk 2]
```diff
@@ -9,23 +9,78 @@
...
+    one_week[row[:tutorial]] = index + 1
+  end
+
+  two_weeks = {}
+  PEGASUS_REPORTING_DB_READONLY.fetch(
+    "SELECT tutorial, #{WEIGHTED_COUNT} #{generate_from_where(end_date - 2.weeks, end_date: end_date, tutorial_pixel: false)} GROUP BY tutorial ORDER BY count DESC"
+  ).each_with_index do |row, index|
+    next if row[:tutorial].nil_or_empty?
+    two_weeks[row[:tutorial]] = index + 1
+  end
```
Comment (ashercodeorg) [inline lib/analyze_hoc_activity_helper.rb]: Any reason these queries can't be combined with a loop? E.g., 

```
[1, 3, 7, 14].each do |num_days|
  start_date = end_date - num_days.days
  ...
end
```

File: lib/analyze_hoc_activity_helper.rb [hunk 3]
```diff
@@ -9,23 +9,78 @@
 DASHBOARD_REPORTING_DB_READONLY = sequel_connect(CDO.dashboard_reporting_db_reader, CDO.dashboard_reporting_db_reader)
 PEGASUS_REPORTING_DB_READONLY = sequel_connect(CDO.pegasus_reporting_db_reader, CDO.pegasus_reporting_db_reader)
 
+WEIGHTED_COUNT = "SUM(" \
+  " IF(session REGEXP '^_.*_'," +
+  # Parses the session weight 'xyz' from the session '_xyz_abcdefghijk'.
+  "   SUBSTRING(SUBSTRING_INDEX(session, '_', 2) FROM 2)," +
+  # The session does not specify the session weight, default to one.
+  "   1)" \
+  " ) AS count"
```
Comment (ashercodeorg) [inline lib/analyze_hoc_activity_helper.rb]: nit: Freeze constants via `.freeze`.

**Comment by Hamms** [comment]
That is indeed the point!

The plan is for Tanya to run this script once a day or so for the couple weeks surrounding hour of code to grab the latest rankings and integrate them into the gsheet.

File: bin/oneoff/rank_tutorials.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env ruby
+#
```
Comment (ashercodeorg) [inline bin/oneoff/rank_tutorials.rb:6]: Any chance you could add a file comment explaining what the script does?

**Pull request merged**